pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -16,143 +16,602 @@
+      // Consume for the next _kafkaReadTime ms, or we get to final offset, whichever happens earlier,
+      // Update _currentOffset upon return from this method
+      Iterable<MessageAndOffset> messagesAndOffsets = _consumerWrapper.fetchMessages(_currentOffset, _endOffset, KAFKA_MAX_FETCH_TIME_MILLIS);
+      Iterator<MessageAndOffset> msgIterator = messagesAndOffsets.iterator();
+
+      int batchSize = 0;
+      while (!_receivedStop && !endCriteriaReached() && msgIterator.hasNext()) {
+        // Get a batch of messages from Kafka
+        // Index each message
+        MessageAndOffset messageAndOffset = msgIterator.next();
+        byte[] array = messageAndOffset.message().payload().array();
+        int offset = messageAndOffset.message().payload().arrayOffset();
+        int length = messageAndOffset.message().payloadSize();
+        GenericRow row = _messageDecoder.decode(array, offset, length);
+
+        if (row != null) {
+          row = _fieldExtractor.transform(row);
+          boolean canTakeMore = _realtimeSegment.index(row);  // Ignore the boolean return
+          if (!canTakeMore) {
+            segmentLogger.warn("We got full during indexing");
+          }
+          batchSize++;
+        }
+        _currentOffset = messageAndOffset.nextOffset();
+        _numRowsConsumed++;
+      }
+      if (batchSize != 0) {
+        segmentLogger.debug("Indexed {} messages current offset {}", batchSize, _currentOffset);
+      } else {
+        // If there were no messages to be fetched from Kafka, wait for a little bit as to avoid hammering the
+        // Kafka broker
+        Uninterruptibles.sleepUninterruptibly(100, TimeUnit.MILLISECONDS);
+      }
+    }
+  }
+
+  public class PartitionConsumer implements Runnable {
+    public void run() {
+      _startTimeMs = now();
+      try {
+        while (!_state.isFinal()) {
+          if (_state.shouldConsume()) {
+            consumeLoop();  // Consume until we reached the end criteria, or we are stopped.
+          }
+          if (_receivedStop) {
+            break;
+          }
+          // If we are sending segmentConsumed() to the controller, we are in HOLDING state.
+          _state = State.HOLDING;
+          SegmentCompletionProtocol.Response response = postSegmentConsumedMsg();
+          SegmentCompletionProtocol.ControllerResponseStatus status = response.getStatus();
+          long rspOffset = response.getOffset();
+          boolean success;
+          switch (status) {
+            case NOT_LEADER:
+              // Retain the same state
+              segmentLogger.warn("Got not leader response");
+              hold();
+              break;
+            case CATCH_UP:
+              if (rspOffset <= _currentOffset) {
+                // Something wrong with the controller. Back off and try again.
+                segmentLogger.error("Invalid catchup offset {} in controller response, current offset {}", rspOffset,
```

Comment:
We cannot download yet, because we don't know whether the segment is committed or not. The best we can do is hold() and go back to the controller who will hopefully get smarter and not ask us to go back in kafka offset. I will be glad to discuss this with you in person if needed.